build-tools/build-indexes

```diff
@@ -637,11 +640,34 @@ process.stdout.write("Building `data/teambuilder-tables.js`... ");
 			const gens = learnset[moveid].map(x => Number(x[0]));
 			const minGen = Math.min(...gens);
 			const vcOnly = minGen === 7 && learnset[moveid].every(x => x[0] !== '7' || x === '7V');
-			if (minGen === 7) learnsets[id][moveid] = '7';
-			if (minGen === 6) learnsets[id][moveid] = '67';
-			if (minGen === 5) learnsets[id][moveid] = '567';
-			if (minGen === 4) learnsets[id][moveid] = '4567';
-			if (minGen === 3) learnsets[id][moveid] = '34567';
+
+			if (gen3HMs.has(moveid) || gen4HMs.has(moveid)) {
+				let legalGens = '';
+				let availableInPrevGen = false;
+				let isHMInPrevGen = false;
+				for (let gen = minGen; gen <= 7; gen++) {
+					if (gens.includes(gen)) {
+						legalGens += gen;
+						availableInPrevGen = true;
+					} else {
+						if (isHMInPrevGen || !availableInPrevGen) {
+							availableInPrevGen = false;
+						} else {
+							legalGens += gen;
+							availableInPrevGen = true;
+						}
+					}
+					isHMInPrevGen = (gen === 3 && gen3HMs.has(moveid)) || (gen === 4 &&  gen4HMs.has(moveid));
+				}
```

Comment:
`availableInPrevGen` and `isHMInPrevGen` are pretty redundant. Here's my refactor of this loop.

```js
				let available = false;
				for (let gen = minGen; gen <= 7; gen++) {
					if (available || gens.includes(gen)) {
						legalGens += gen;
						available = true;
					}
					if (gen === 3 && gen3HMs.has(moveid) || gen === 4 && gen4HMs.has(moveid)) {
						available = false;
					}
				}
```

Which actually can be further unrolled:

```js
				let legalGens = '';
				let available = false;
				if (minGen === 3) {
					legalGens += '3';
					available = !gen3HMs.has(moveid);
				}
				if (available || gens.includes(4)) {
					legalGens += 4;
					available = !gen4HMs.has(moveid);
				}
				let minUpperGen = available ? 5 : Math.min(
					...gens.filter(gen => gen > 4)
				);
				legalGens += '01234567'.slice(minUpperGen);
```

Reply:
n.b. `minUpperGen` can be `Infinity`, but this code will still work correctly if it is.